pkg/object/httpserver/spec.go

```diff
@@ -136,8 +136,16 @@ func (spec *Spec) tlsConfig() (*tls.Config, error) {
 		Certificates: certificates,
 		NextProtos:   []string{"acme-tls/1"},
 	}
-	tlsConf.GetCertificate = func(chi *tls.ClientHelloInfo) (*tls.Certificate, error) {
-		return autocertmanager.GetCertificate(chi, !spec.AutoCert /* tokenOnly */)
+
+	if spec.AutoCert {
+		acm, exists := autocertmanager.GetGlobalAutoCertManager()
+		if !exists {
+			return nil, fmt.Errorf("BUG: autocert manager is not initialized")
+		}
+
+		tlsConf.GetCertificate = func(chi *tls.ClientHelloInfo) (*tls.Certificate, error) {
+			return acm.GetCertificate(chi, !spec.AutoCert /* tokenOnly */)
+		}
```

Comment:
No need to use atomic.Value because it is called infrequently. But you're right we need to get acm in real time in `tlsConf.GetCertificate`.

Reply:
`tlsConf. GetCertificate ` is used to get certificate based on client hello message. When a client connect to `https` server, they send a client hello message. And golang use this client hello info to get certificate from autocertmanager. I am not sure if golang use cache for this. If not, it will be called every time a client connect to server. That can be a lot...

Anyway, `atomic.Value` brings no harm at all. Only benefits of efficiency.

Reply:
Updated